lightning/src/chain/chainmonitor.rs

```diff
@@ -0,0 +1,226 @@
+//! setup to monitor channels remotely. In the latter case, a custom `chain::Watch` implementation
+//! would be responsible for routing each update to a remote server and for retrieving monitor
+//! events. The remote server would make use of `ChainMonitor` for block processing and for
+//! servicing `ChannelMonitor` updates from the client.
+//!
+//! [`ChainMonitor`]: struct.ChainMonitor.html
+//! [`chain::Filter`]: ../trait.Filter.html
+//! [`chain::Watch`]: ../trait.Watch.html
+//! [`ChannelMonitor`]: ../channelmonitor/struct.ChannelMonitor.html
+//! [`MonitorEvent`]: ../channelmonitor/enum.MonitorEvent.html
+
+use bitcoin::blockdata::block::BlockHeader;
+
+use chain;
+use chain::Filter;
+use chain::chaininterface::{BroadcasterInterface, FeeEstimator};
+use chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr, MonitorEvent, MonitorUpdateError};
+use chain::transaction::{OutPoint, TransactionData};
+use chain::keysinterface::ChannelKeys;
+use util::logger::Logger;
+use util::events;
+use util::events::Event;
+
+use std::collections::{HashMap, hash_map};
+use std::sync::Mutex;
+use std::ops::Deref;
+
+/// An implementation of [`chain::Watch`] for monitoring channels.
+///
+/// Connected and disconnected blocks must be provided to `ChainMonitor` as documented by
+/// [`chain::Watch`]. May be used in conjunction with [`ChannelManager`] to monitor channels locally
+/// or used independently to monitor channels remotely. See the [module-level documentation] for
+/// details.
+///
+/// [`chain::Watch`]: ../trait.Watch.html
+/// [`ChannelManager`]: ../../ln/channelmanager/struct.ChannelManager.html
+/// [module-level documentation]: index.html
+pub struct ChainMonitor<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref>
+	where C::Target: chain::Filter,
+        T::Target: BroadcasterInterface,
+        F::Target: FeeEstimator,
+        L::Target: Logger,
+{
+	/// The monitors
+	pub monitors: Mutex<HashMap<OutPoint, ChannelMonitor<ChanSigner>>>,
+	chain_source: Option<C>,
+	broadcaster: T,
+	logger: L,
+	fee_estimator: F
+}
+
+impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonitor<ChanSigner, C, T, F, L>
+	where C::Target: chain::Filter,
+	      T::Target: BroadcasterInterface,
+	      F::Target: FeeEstimator,
+	      L::Target: Logger,
+{
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel and reacting accordingly based on transactions in the connected block. See
+	/// [`ChannelMonitor::block_connected`] for details. Any HTLCs that were resolved on chain will
+	/// be returned by [`chain::Watch::release_pending_monitor_events`].
+	///
+	/// Calls back to [`chain::Filter`] if any monitor indicated new outputs to watch, returning
+	/// `true` if so. In this case, if providing pre-filtered blocks, the caller should re-fetch the
+	/// block to obtain updated `txdata` and re-call `block_connected`.
```

Comment:
> Luckily this is no longer the way we're going to go (yay tons less complexity)...that said we need to note _somewhere_ that if you do filter a block you must also include any in-block descendants of matched transactions...I presume in the docs for Filter?

Updated `chain::Filter` docs to mention in-block descendants of matched transactions

> Probably means we can simplify the code, too.

By this I take it you mean we no longer need to return a bool and have the caller refetch/rescan? I tried removing the loop from `connect_block` in our functional test utils, but unfortunately doing so causes six tests to fail.

```
failures:
    ln::functional_tests::test_claim_on_remote_revoked_sizeable_push_msat
    ln::functional_tests::test_claim_on_remote_sizeable_push_msat
    ln::functional_tests::test_dynamic_spendable_outputs_local_htlc_timeout_tx
    ln::functional_tests::test_key_derivation_params
    ln::functional_tests::test_static_spendable_outputs_justice_tx_revoked_htlc_success_tx
    ln::functional_tests::test_static_spendable_outputs_timeout_tx
```

Is this an indication that the blocks used in the tests need to include descendant transactions?

Reply:
Hmm, I can take a look but I'd bet that they're simply assuming reorgs which double-generate transactions needlessly. To get this over the line I'm happy to ignore that for now, but we should fix it before release.

Reply:
We should still update the docs to note that you do not need to do this (even if the tests do because they're overly restrictive), and we *should* note that you have to include descendants of matched transactions.

Reply:
Leaving simplification for a follow-up. Re-worded `ChainMonitor::block_connected` docs. The part about not needing to refetch seems a little out of place. Let me know if you have a better suggestion.